File: DSA/DSA/LinkedList/LList.py
class Node: 
# CREATING VALUES, RETURNING HEAD, SHOWVALUES. 
    def __init__(self, data, next= None):
        self.data= data
        self.next= next
      
     
    def head(list):
        if len(list)== 0:                   # smallest edge case.
            return None
        for i in range(len(list)):           
            if i==0:
                h= Node(list[i])            
                p=h                         #No need to pass the id(Node(list[i]))
                
            else:
                c= Node(list[i])
                p.next= c
                p=c
        return h                             
    
    
    def showValues(head):
        temp= head
        if temp== None: 
            return None
        while(temp): 
            print(temp.data, end= " ")
            temp= temp.next
        return        
        
        
# DELETING HEAD. 
    def deletingHead(head):
        temp= head
        if temp== None: 
            temp= head
            head= head.next
            del temp
        return head


# DELETING TAIL.
    def deletingTail1(head):
        temp= head
        if (temp == None or temp.next== None):      #edge case.  
            return None
        last= None 
        while(temp):                                #before didn't include the "last."
            secondlast= last 
            last= temp
            temp = temp.next
            
        del last       
        return secondlast.data
            
        
    def deletingTail2(head): 
        temp= head
        if (temp == None or temp.next== None):      #edge case.  
            return None
        while(temp.next.next != None): 
            temp= temp.next
        del temp.next
        return temp.data

    
    def deletingKthElement(head, position): 
        if (head== None):
            return None 
        temp= head
        count=1
        while(temp):
            if( position==1): 
                head= temp.next
                del temp
                
            
            
            elif(temp.next==None):  # no need to make a  case for last element.
                if (position==count): 
                    del temp.next
                    temp.next= None 
                    break
        
        
            else:                           #this is taking care of the elif block. 
                if(count== position-1): 
                    var= temp.next.next
                    del temp.next
                    temp.next= var
                    break
                
            temp=temp.next
            count+=1
        return head   
   
         
    
    def deletingAValue(head, val):              
        temp= head
        pervious= None
        current= None
        while(temp):
            
            if (current != None): 
                pervious=  current     #second and secondlast approach
            current = temp 
            
            if (temp.data== val):
                
                pervious.next= current.next
                del current 
                temp= pervious
                current= pervious 
                
            temp= temp.next
    
        return head
        
        
    def insertThisValue(head, position, val):
        count = 1
        temp = head
         
        while temp:
            previous = current
            current = temp
            if count == 1:
                x = Node(val)
                x.next = temp
                break
            
            else:  # Indentation fixed here
                if count == position:
                        x = Node(val)
                        x.next = current
                        previous.next = x
                        

    
        
            
    @staticmethod   
    def main(): 
       list=[1,2,3,4,5,6,7,8,9,10] 
       head= Node.head(list)    
       head= Node.test(head)
       Node.showValues(head)
      
       
Node.main()

File: DSA/DSA/LinkedList/LList.cpp
l#include<bits/stdc++.h> 
using namespace std; 


// Creating values, showvalues. 
class Node{
    public:                                          // 1. so we can use this outside the Node class.
    int data; 
    Node* next;                                
            

    
    public: 
    Node(int data1, Node* location= nullptr ) {       // In taking the values, using Node. 
        data= data1; 
        next= location; 
    }

}; 


Node* head(){
    int n, val; 
    cin >> n; 
    cin >> val; 
    Node* x= new Node(val); 
    Node* h= x ; 
    Node* p= x ;                                    // 2. Must outside the function if we don't then p and h remains undefined. 
    for (int i=1; i<=n-1; i++){
        cin >> val; 
        Node* c = new Node(val);                    // 3. "new" is must. 
        p->next = c; 
        p = c ; 
    }  
    return h; 
}


auto showValues(Node* head){
    Node* temp = head; 
    while(temp){
        cout << temp-> data << " "; 
        temp=temp->next; 
    }
    
}
Node* deletingHead(Node* head){
    if(head == nullptr) return head ;                                // when list is empty. 
    Node* temp= head; 
    head= temp->next; 
    delete temp; 
    return head; 
}


int deletingTail(Node* head){
    Node* secondlast= head;                         // 1. last and secondlast approach. 
    Node*last= head; 
    Node* temp= head; 
    while(temp){
        secondlast= last; 
        last= temp;
        temp= temp->next; 
    }
    delete last; 
    return secondlast->data; 
}


Node* deletingTail2(Node* head ){
    Node* temp = head; 
    Node* x; 
    if( temp == NULL| temp->next == NULL) return NULL;   
    while(temp->next->next!=nullptr ){// 2. remember temp->next-> next.  
        temp = temp->next; 
    }
    delete temp->next; 
    temp-> next= nullptr; 
    return head ; 
}


// insertion of an Element. 
Node* insertAnElement(Node* head,int  val){
    Node* temp = head; 
    Node* x= new Node(val); 
    while(temp->next != nullptr){               // we can use nullp
        temp= temp->next; 
    }
    temp->next= x; 
    return head; 
}
    
int main(){
    Node* temp= head();
    Node* head= deletingTail2(temp );   
    showValues(head); 
   
    return 0; 
}

